Sources/swift-format/Frontend/Frontend.swift

```diff
@@ -16,6 +16,118 @@ import SwiftParser
 import SwiftSyntax
 
 class Frontend {
+  /// Provides formatter configurations for given `.swift` source files, configuration files or configuration strings.
+  struct ConfigurationProvider {
```

Comment:
Just wondering: Would it be simpler if we just had a static method (or global function) like `func configuration(forConfigPathOrString pathOrString: String?, orForSwiftFileAt swiftFileURL: URL?, diagnosticsEngine: DiagnosticsEngine)` and then either make `checkForUnrecognizedRules` a nested function or make it take a `DiagnosticsEngine` as well?  To me, having a type here implies that there is some state being stored, which isn’t actually the case.

Reply:
> To me, having a type here implies that there is some state being stored, which isn’t actually the case.

There is state being stored in the form of `configurationLoader`, which stores the cache for previously loaded configs.

Reply:
Oh, I see. Fair enough then.